pallets/api/src/messaging/transports/ismp.rs

```diff
@@ -164,90 +156,120 @@ impl<T: Config> IsmpModule for Handler<T> {
-	let (origin, id) =
-		IsmpRequests::<T>::get(commitment).ok_or(Error::Custom("request not found".into()))?;
+	ensure!(
+		response_data.encoded_size() <= T::MaxResponseLen::get() as usize,
+		Error::Custom("Response length exceeds maximum allowed length.".into())
+	);
 
-	let Some(super::super::Message::Ismp { commitment, callback, deposit }) =
-		Messages::<T>::get(&origin, &id)
+	let (initiating_origin, id) =
+		IsmpRequests::<T>::get(commitment).ok_or(Error::Custom("Request not found.".into()))?;
+
+	let Some(super::super::Message::Ismp { commitment, callback, message_deposit }) =
+		Messages::<T>::get(&initiating_origin, id)
 	else {
-		return Err(Error::Custom("message not found".into()))?
+		return Err(Error::Custom("Message must be an ismp request.".into()).into());
 	};
 
+	// Deposit that the message has been recieved before a potential callback execution.
+	Pallet::<T>::deposit_event(event(initiating_origin.clone(), id));
+
 	// Attempt callback with result if specified.
 	if let Some(callback) = callback {
-		// TODO: check response length
-		if Pallet::<T>::call(origin.clone(), callback, id, &encode, deposit).is_ok() {
-			Pallet::<T>::deposit_event(event(origin, id));
+		if Pallet::<T>::call(&initiating_origin, callback, &id, response_data).is_ok() {
+			// Clean storage, return deposit
+			Messages::<T>::remove(&initiating_origin, id);
+			IsmpRequests::<T>::remove(commitment);
+			T::Fungibles::release(
+				&HoldReason::Messaging.into(),
+				&initiating_origin,
+				message_deposit,
+				Exact,
+			)
+			.map_err(|_| Error::Custom("failed to release message deposit.".into()))?;
+
 			return Ok(());
 		}
 	}
 
-	// Otherwise store response for manual retrieval and removal.
-	let response: BoundedVec<u8, T::MaxResponseLen> =
-		store().try_into().map_err(|_| Error::Custom("response exceeds max".into()))?;
+	// No callback or callback error: store response for manual retrieval and removal.
+	let encoded_response: BoundedVec<u8, T::MaxResponseLen> = response_data
+		.encode()
+		.try_into()
+		.map_err(|_| Error::Custom("response exceeds max".into()))?;
 	Messages::<T>::insert(
-		&origin,
-		&id,
-		super::super::Message::IsmpResponse { commitment, deposit, response },
+		&initiating_origin,
+		id,
+		super::super::Message::IsmpResponse {
+			commitment,
+			message_deposit,
+			response: encoded_response,
+		},
 	);
-	Pallet::<T>::deposit_event(event(origin, id));
+	Ok(())
+}
+
+pub(crate) fn timeout_commitment<T: Config>(commitment: &H256) -> Result<(), anyhow::Error> {
+	let key = IsmpRequests::<T>::get(commitment)
+		.ok_or(Error::Custom("Request commitment not found while processing timeout.".into()))?;
+	Messages::<T>::try_mutate(key.0, key.1, |message| {
+		let Some(super::super::Message::Ismp { commitment, message_deposit, callback }) = message
+		else {
+			return Err(Error::Custom("Invalid message".into()));
+		};
+		let callback_deposit = callback.map(|cb| T::WeightToFee::weight_to_fee(&cb.weight));
```

Comment:
Would prefer this is only calculated at time of deposit taken, as mentioned elsewhere.